clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java

```diff
@@ -0,0 +1,366 @@
+        }
+
+        // Update schemas
+        SchemaData schemaData = new SchemaData(storeName);
+        for (Map.Entry<Integer, String> entry: keySchema.entrySet()) {
+          schemaData.setKeySchema(new SchemaEntry(entry.getKey(), entry.getValue()));
+        }
+        for (Map.Entry<Integer, String> entry: valueSchemas.entrySet()) {
+          schemaData.addValueSchema(new SchemaEntry(entry.getKey(), entry.getValue()));
+        }
+        schemas.set(schemaData);
+
+        // Evict entries that are over two versions behind from the latest fetched version
+        readyToServeInstancesMap.entrySet()
+            .removeIf(
+                entry -> Integer.parseInt(entry.getKey().split(VERSION_PARTITION_SEPARATOR)[0]) < fetchedVersion - 2);
+        versionPartitionerMap.entrySet().removeIf(entry -> entry.getKey() < fetchedVersion - 2);
+        versionZstdDictionaryMap.entrySet().removeIf(entry -> entry.getKey() < fetchedVersion - 2);
+
+        // Wait for dictionary fetch to finish if there is one
+        try {
+          if (dictionaryFetchFuture != null) {
+            dictionaryFetchFuture.get(ZSTD_DICT_FETCH_TIMEOUT, TimeUnit.SECONDS);
+          }
+          currentVersion.set(fetchedVersion);
+          clusterStats.updateCurrentVersion(getCurrentStoreVersion());
+          latestSuperSetValueSchemaId.set(newSuperSetValueSchemaId);
+        } catch (InterruptedException interruptedException) {
+          Thread.currentThread().interrupt();
+          throw new VeniceClientException("Dictionary fetch operation was interrupted");
+        } catch (ExecutionException | TimeoutException e) {
+          LOGGER.warn(
+              "Dictionary fetch operation could not complete in time for some of the versions. "
+                  + "Will be retried on next refresh",
+              e);
+          clusterStats.recordVersionUpdateFailure();
+        }
+      }
+    } catch (InterruptedException interruptedException) {
+      Thread.currentThread().interrupt();
+      throw new VeniceClientException("Metadata fetch operation was interrupted");
+    } catch (ExecutionException e) {
+      // perform an on demand refresh if update fails in case of store migration
+      // TODO: need a better way to handle store migration
+      if (!onDemandRefresh) {
+        LOGGER.warn("Metadata fetch operation has failed with exception {}", e.getMessage());
+        discoverD2Service(true);
+        updateCache(true);
+      } else {
+        // pass the error along if the on demand refresh also fails
+        throw new VeniceClientException("Metadata fetch retry has failed", e.getCause());
+      }
+    }
+
+    // update the refresh interval to a random interval and queue the next refresh
+    long randomRefreshInterval = refreshIntervalInSeconds + ThreadLocalRandom.current().nextInt(-10, 10);
+    scheduler.schedule(this::refresh, randomRefreshInterval, TimeUnit.SECONDS);
```

Comment:
It makes to move to `start()` but I'm wondering why schedule at a fixed rate. I thought we wanted some randomness in the interval to avoid a refresh storm?

Reply:
I think the number of concurrent requests will be insignificant. Since it also requires users to deploy their application around the same time (which is unlikely). If you'd like to have some variation it's fine too, but we need to add it into a finally block as you mentioned to prevent random exceptions putting the client to never update metadata again.

Reply:
Keeping it simple and scheduling from `start()` without any randomness